taichi/program/callable.h

```diff
@@ -16,22 +16,45 @@ class TI_DLL_EXPORT Callable {
   std::unique_ptr<FrontendContext> context{nullptr};
 
   struct Arg {
-    DataType dt;
     bool is_array{
         false};  // This is true for both ndarray and external array args.
-    std::size_t total_dim{0};             // total dim of array
-    std::vector<int> element_shape = {};  // shape of each element
+    std::size_t total_dim{0};  // total dim of array
 
     explicit Arg(const DataType &dt = PrimitiveType::unknown,
                  bool is_array = false,
                  std::size_t size_unused = 0,
                  int total_dim = 0,
-                 std::vector<int> element_shape = {})
-        : dt(dt),
-          is_array(is_array),
-          total_dim(total_dim),
-          element_shape(std::move(element_shape)) {
+                 std::vector<int> element_shape = {}) {
+      if (dt->is<PrimitiveType>() && element_shape.size() > 0) {
+        this->dt_ = taichi::lang::TypeFactory::get_instance().get_tensor_type(
+            element_shape, dt.operator->());
+
+      } else {
+        this->dt_ = dt;
+      }
+
+      this->is_array = is_array;
+      this->total_dim = total_dim;
+    }
+
+    std::vector<int> get_element_shape() const {
+      return dt_.get_shape();
+    }
+
+    DataType get_element_type() const {
+      return dt_.get_element_type();
+    }
+
+    int get_element_size() const {
+      return data_type_size(dt_);
     }
+
+    DataType get_dtype() const {
+      return dt_;
+    }
+
+   private:
```

Comment:
nit: shall we add a note here (or near DataType class) to explain its difference with primitive types as you put in the comment above? 
That way it's easier to ppl who read the codebase. ;)

Reply:
comments added~